lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs

```diff
@@ -163,152 +179,250 @@ parseGenesisData = \case
-data CardanoNodeConfig = CardanoNodeConfig
-    { nodeConfigFile   :: FilePath
-    , nodeDatabaseDir  :: FilePath
-    , nodeSocketFile   :: FilePath
-    , nodeTopologyFile :: FilePath
-    , nodeVrfKey       :: FilePath
-    , nodeKesKey       :: FilePath
-    , nodeOpCert       :: FilePath
-    }
-
--- | Spins up a @cardano-node@ in another process.
+-- | A quick helper to interact with the 'cardano-cli'. Assumes the cardano-cli
+-- is available in PATH.
+cli :: [String] -> IO String
+cli args =
+    readProcess "cardano-cli" args stdin
+  where
+    stdin = ""
+
+-- | Execute an action after starting a cluster of stake pools. The cluster also
+-- contains a single BFT node that is pre-configured with keys available in the
+-- test data.
 --
--- IMPORTANT: @cardano-node@ must be available on the current path.
-withCardanoNode
+-- This BFT node is essential in order to bootstrap the chain and allow
+-- registering pools. Passing `0` as a number of pool will simply start a single
+-- BFT node.
+withCluster
     :: Trace IO Text
-    -- ^ Some trace for logging
-    -> FilePath
-    -- ^ Test directory
+    -- ^ Trace for subprocess control logging
     -> Severity
+    -- ^ Minimal logging severity
+    -> Int
+    -- ^ How many pools should the cluster spawn.
     -> (FilePath -> Block -> (NetworkParameters, NodeVersionData) -> IO a)
-    -- ^ Callback function with a socket description and genesis params
-    -> IO a
-withCardanoNode tr tdir severity action =
-    orThrow $ withConfig tdir severity $ \cfg block0 (np, vData) -> do
-        nodePort <- getRandomPort
-        let args = mkArgs cfg nodePort
+    -- ^ Action to run with the cluster up
+    -> IO (Either ProcessHasExited a)
+withCluster tr severity n action = do
+    ports <- randomUnusedTCPPorts (n + 1)
+    print $ permutations ports
+    withBFTNode tr severity (head $ permutations ports) $ \socket block0 params -> do
+        waitGroup <- newChan
+        doneGroup <- newChan
+        let waitAll   = replicateM  n (readChan waitGroup)
+        let cancelAll = replicateM_ n (writeChan doneGroup ())
+
+        forM_ (init $ permutations ports) $ \(port, peers) -> do
+            withAsync
+                (withStakePool tr severity (port, peers) $ readChan doneGroup)
+                (writeChan waitGroup . Just) -- FIXME: register pool
+              `finally` (writeChan waitGroup Nothing)
+
+        group <- waitAll
+        if length (catMaybes group) /= n then do
+            cancelAll
+            throwIO $ ProcessHasExited
+                "cluster didn't start correctly"
+                (ExitFailure 1)
+        else do
+            action socket block0 params `finally` cancelAll
+  where
+    -- | Get permutations of the size (n-1) for a list of n elements, alongside with
+    -- the element left aside. `[a]` is really expected to be `Set a`.
+    --
+    -- >>> permutations [1,2,3]
+    -- [(1,[2,3]), (2, [1,3]), (3, [1,2])]
+    permutations :: Eq a => [a] -> [(a, [a])]
```

Comment:
maybe reuse https://hackage.haskell.org/package/base-4.14.0.0/docs/src/Data.OldList.html#permutations

Reply:
and then use `f(x:xs)=[x,[xs]]` ?

Reply:
It won't work because it's does not actually compute _all_ permutations (unlike Data.List.permutations). Using Data.List.permutations on the example above, we would also get [(1, [3,2]), (2, [3,1]), (3, [2,1])]. So, we should also nub + sort the second element which is arguably simpler than the current function :thinking: .. 

Changed for:

```hs
    rotate :: Ord a => [a] -> [(a, [a])]
    rotate = nub . fmap (\(x:xs) -> (x, sort xs)) . permutations
```